chalk-solve/src/infer/unify.rs

```diff
@@ -660,7 +740,9 @@ impl<'t, I: Interner> Unifier<'t, I> {
                 if matches!(kind, TyVariableKind::Integer | TyVariableKind::Float) {
                     ty.clone()
                 } else if let Some(ty) = self.table.normalize_ty_shallow(interner, ty) {
-                    self.generalize_ty(&ty, universe_index)
+                    self.generalize_ty(&ty, universe_index, variance)
+                } else if matches!(variance, Variance::Invariant) {
```

Comment:
The rustc code here also checks that this variable can name `universe_index` here, although that's maybe only for error messages.

Reply:
Hmm. So should I change this here?

Reply:
I think that we can keep the simpler code for now.